titanic 연습

In [1]:
import pandas as pd
import numpy as np

In [2]:
train =pd.read_csv('./train.csv')
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
train.shape

(891, 12)

In [4]:
train.columns = train.columns.str.lower()
train.columns

Index(['passengerid', 'survived', 'pclass', 'name', 'sex', 'age', 'sibsp',
       'parch', 'ticket', 'fare', 'cabin', 'embarked'],
      dtype='object')

In [5]:
train.isna().sum()[train.isna().sum()>0]

age         177
cabin       687
embarked      2
dtype: int64

출력결과 다음과 같이 출력
col    cnt     nan_cnt     nan_rate
age    891     177         23%

In [6]:
col = train.columns
cnt = train.notna().sum()
nan_cnt = train.isna().sum()
nan_rate = round((nan_cnt/train.shape[0])*100,2 )

dic = {"col":col, "cnt":cnt, "nan_cnt":nan_cnt, "nan_rate":nan_rate}
train_dic = pd.DataFrame(dic)
(train_dic.sort_values('nan_cnt', ascending = False)).T

cabin    age  embarked  passengerid  survived  pclass  name  sex  \
col       cabin    age  embarked  passengerid  survived  pclass  name  sex   
cnt         204    714       889          891       891     891   891  891   
nan_cnt     687    177         2            0         0       0     0    0   
nan_rate   77.1  19.87      0.22          0.0       0.0     0.0   0.0  0.0   

          sibsp  parch  ticket  fare  
col       sibsp  parch  ticket  fare  
cnt         891    891     891   891  
nan_cnt       0      0       0     0  
nan_rate    0.0    0.0     0.0   0.0

In [7]:
train['age'].unique()

array([22.  , 38.  , 26.  , 35.  ,   nan, 54.  ,  2.  , 27.  , 14.  ,
        4.  , 58.  , 20.  , 39.  , 55.  , 31.  , 34.  , 15.  , 28.  ,
        8.  , 19.  , 40.  , 66.  , 42.  , 21.  , 18.  ,  3.  ,  7.  ,
       49.  , 29.  , 65.  , 28.5 ,  5.  , 11.  , 45.  , 17.  , 32.  ,
       16.  , 25.  ,  0.83, 30.  , 33.  , 23.  , 24.  , 46.  , 59.  ,
       71.  , 37.  , 47.  , 14.5 , 70.5 , 32.5 , 12.  ,  9.  , 36.5 ,
       51.  , 55.5 , 40.5 , 44.  ,  1.  , 61.  , 56.  , 50.  , 36.  ,
       45.5 , 20.5 , 62.  , 41.  , 52.  , 63.  , 23.5 ,  0.92, 43.  ,
       60.  , 10.  , 64.  , 13.  , 48.  ,  0.75, 53.  , 57.  , 80.  ,
       70.  , 24.5 ,  6.  ,  0.67, 30.5 ,  0.42, 34.5 , 74.  ])

In [8]:
train['age'].nunique()

88

In [9]:
train['age'].value_counts()

24.00    30
22.00    27
18.00    26
19.00    25
28.00    25
         ..
36.50     1
55.50     1
0.92      1
23.50     1
74.00     1
Name: age, Length: 88, dtype: int64

In [10]:
for col in train.columns:
    if train[col].nunique() < 100:
        print(col, train[col].unique(), train[col].nunique())

survived [0 1] 2
pclass [3 1 2] 3
sex ['male' 'female'] 2
age [22.   38.   26.   35.     nan 54.    2.   27.   14.    4.   58.   20.
 39.   55.   31.   34.   15.   28.    8.   19.   40.   66.   42.   21.
 18.    3.    7.   49.   29.   65.   28.5   5.   11.   45.   17.   32.
 16.   25.    0.83 30.   33.   23.   24.   46.   59.   71.   37.   47.
 14.5  70.5  32.5  12.    9.   36.5  51.   55.5  40.5  44.    1.   61.
 56.   50.   36.   45.5  20.5  62.   41.   52.   63.   23.5   0.92 43.
 60.   10.   64.   13.   48.    0.75 53.   57.   80.   70.   24.5   6.
  0.67 30.5   0.42 34.5  74.  ] 88
sibsp [1 0 3 4 2 5 8] 7
parch [0 1 2 5 3 4 6] 7
embarked ['S' 'C' 'Q' nan] 3


In [11]:
# ! pip install matplotlib

In [12]:
import matplotlib.pyplot as plot

In [13]:
traincp = train.copy()


In [14]:
traincp = traincp.drop('pclass', axis = 1)
traincp.head()

passengerid  survived                                               name  \
0            1         0                            Braund, Mr. Owen Harris   
1            2         1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3         1                             Heikkinen, Miss. Laina   
3            4         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4            5         0                           Allen, Mr. William Henry   

      sex   age  sibsp  parch            ticket     fare cabin embarked  
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S  
1  female  38.0      1      0          PC 17599  71.2833   C85        C  
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  female  35.0      1      0            113803  53.1000  C123        S  
4    male  35.0      0      0            373450   8.0500   NaN        S

In [15]:
del traincp

In [16]:
train.groupby(['sex','survived'])['survived'].count()

sex     survived
female  0            81
        1           233
male    0           468
        1           109
Name: survived, dtype: int64

In [17]:
train[['pclass','survived']].value_counts().sort_index()

pclass  survived
1       0            80
        1           136
2       0            97
        1            87
3       0           372
        1           119
dtype: int64

In [18]:
train['cabin'][train['cabin'].notna()]

1              C85
3             C123
6              E46
10              G6
11            C103
          ...     
871            D35
872    B51 B53 B55
879            C50
887            B42
889           C148
Name: cabin, Length: 204, dtype: object

In [19]:
train['cabin2'] = train['cabin'].str[0:1]
train[['cabin','cabin2']].head(10)

cabin cabin2
0   NaN    NaN
1   C85      C
2   NaN    NaN
3  C123      C
4   NaN    NaN
5   NaN    NaN
6   E46      E
7   NaN    NaN
8   NaN    NaN
9   NaN    NaN

In [20]:
train[['cabin2','survived']].value_counts().sort_index()

cabin2  survived
A       0            8
        1            7
B       0           12
        1           35
C       0           24
        1           35
D       0            8
        1           25
E       0            8
        1           24
F       0            5
        1            8
G       0            2
        1            2
T       0            1
dtype: int64

In [21]:
train[['pclass','cabin2']].value_counts().sort_index()

pclass  cabin2
1       A         15
        B         47
        C         59
        D         29
        E         25
        T          1
2       D          4
        E          4
        F          8
3       E          3
        F          5
        G          4
dtype: int64

In [22]:
train[['embarked','survived']].value_counts().sort_index()

embarked  survived
C         0            75
          1            93
Q         0            47
          1            30
S         0           427
          1           217
dtype: int64

In [23]:
train['embarked'].isna().sum()

2

In [24]:
mode_val = train['embarked'].mode().values[0]
mode_val

'S'

In [25]:
train['embarked'] = train['embarked'].fillna(mode_val)
train['embarked']

0      S
1      C
2      S
3      S
4      S
      ..
886    S
887    S
888    S
889    C
890    Q
Name: embarked, Length: 891, dtype: object

In [26]:
train['embarked'].isna().sum()

0

In [27]:
train[['sex','survived']].value_counts().sort_index()

sex     survived
female  0            81
        1           233
male    0           468
        1           109
dtype: int64

In [28]:
train['name'].head(10)

0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
5                                     Moran, Mr. James
6                              McCarthy, Mr. Timothy J
7                       Palsson, Master. Gosta Leonard
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
9                  Nasser, Mrs. Nicholas (Adele Achem)
Name: name, dtype: object

In [29]:
train['name_title'] = train['name'].str.extract('\s([\w]+)[.]\s')
train['name_title'].value_counts()

Mr          517
Miss        182
Mrs         125
Master       40
Dr            7
Rev           6
Mlle          2
Major         2
Col           2
Countess      1
Capt          1
Ms            1
Sir           1
Lady          1
Mme           1
Don           1
Jonkheer      1
Name: name_title, dtype: int64

In [30]:
traincp = train.copy()
traincp.head()

passengerid  survived  pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                name     sex   age  sibsp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   parch            ticket     fare cabin embarked cabin2 name_title  
0      0         A/5 21171   7.2500   NaN        S    NaN         Mr  
1      0          PC 17599  71.2833   C85        C      C        Mrs  
2      0  STON/O2. 3101282   7.9250   NaN        S    NaN       Miss  
3      0            113803  53.1000  C123        S      C        Mrs  
4      0            373450   8.0500   NaN        S    NaN         Mr

In [31]:
traincp['pclass'].replace(3,33)

0      33
1       1
2      33
3       1
4      33
       ..
886     2
887     1
888    33
889     1
890    33
Name: pclass, Length: 891, dtype: int64

In [32]:
traincp['pclass'].replace([1,2,3],[11,22,333])

0      333
1       11
2      333
3       11
4      333
      ... 
886     22
887     11
888    333
889     11
890    333
Name: pclass, Length: 891, dtype: int64

In [33]:
dic = {"female":0,"male":1}
traincp['sex'] = traincp['sex'].map(dic)
traincp['sex']

0      1
1      0
2      0
3      0
4      1
      ..
886    1
887    0
888    0
889    1
890    1
Name: sex, Length: 891, dtype: int64

In [34]:
traincp['sex'] = traincp['sex'].apply(lambda x: 'male' if x == 1 else 'female')
traincp['sex']

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: sex, Length: 891, dtype: object

In [35]:
traincp['name_title2'] = traincp['name_title']
traincp['name_title2'] = traincp['name_title2'].replace(['Rev','Major','Major','Col','Capt','Sir','Don','Jonkheer'],['Mr','Mr','Mr','Mr','Mr','Mr','Mr','Mr'])
traincp['name_title2'] = traincp['name_title2'].replace(['Mlle','Ms','Lady'],['Miss','Miss','Miss'])
traincp['name_title2'] = traincp['name_title2'].replace(['Countess','Mme'],['Mrs','Mrs'])
traincp['name_title2'].unique()

array(['Mr', 'Mrs', 'Miss', 'Master', 'Dr'], dtype=object)

In [36]:
train['age'].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: age, dtype: float64

In [37]:
train['age'].isna().sum()

177

In [38]:
train['name_title2'] = train['name_title']
train['name_title2'] = train['name_title2'].replace(['Rev','Major','Major','Col','Capt','Sir','Don','Jonkheer'],
                              ['Mr','Mr','Mr','Mr','Mr','Mr','Mr','Mr'])
train['name_title2'] = train['name_title2'].replace(['Mlle','Ms','Lady'],['Miss','Miss','Miss'])
train['name_title2'] = train['name_title2'].replace(['Countess','Mme'],['Mrs','Mrs'])
train['name_title2'].value_counts()

Mr        531
Miss      186
Mrs       127
Master     40
Dr          7
Name: name_title2, dtype: int64

In [39]:
train[train['age'].notna()].groupby('name_title2')['age'].mean()

name_title2
Dr        42.000000
Master     4.574167
Miss      22.020000
Mr        32.891990
Mrs       35.763636
Name: age, dtype: float64

In [40]:
def age_minus(x):
    return x - 1

# traincp.groupby('name_title2')['age'].apply(age_minus)  
train.groupby('name_title2', group_keys=True)['age'].apply(age_minus)


name_title2     
Dr           245    43.0
             317    53.0
             398    22.0
             632    31.0
             660    49.0
                    ... 
Mrs          871    46.0
             874    27.0
             879    55.0
             880    24.0
             885    38.0
Name: age, Length: 891, dtype: float64

In [41]:
train.groupby('name_title2', group_keys=True)['age'].transform(age_minus)

0      21.0
1      37.0
2      25.0
3      34.0
4      34.0
       ... 
886    26.0
887    18.0
888     NaN
889    25.0
890    31.0
Name: age, Length: 891, dtype: float64

In [42]:
traincp.groupby('name_title2').apply(lambda gx : age_minus(gx['age']))

name_title2     
Dr           245    43.0
             317    53.0
             398    22.0
             632    31.0
             660    49.0
                    ... 
Mrs          871    46.0
             874    27.0
             879    55.0
             880    24.0
             885    38.0
Name: age, Length: 891, dtype: float64

In [43]:
train.groupby('name_title2')['age'].mean()

name_title2
Dr        42.000000
Master     4.574167
Miss      22.020000
Mr        32.891990
Mrs       35.763636
Name: age, dtype: float64

In [44]:
train['age2'] = train.groupby('name_title2')['age'].transform(lambda g: g.fillna(g.mean()))
train[['name_title2','age','age2']][train['age'].isna()]

name_title2  age       age2
5            Mr  NaN  32.891990
17           Mr  NaN  32.891990
19          Mrs  NaN  35.763636
26           Mr  NaN  32.891990
28         Miss  NaN  22.020000
..          ...  ...        ...
859          Mr  NaN  32.891990
863        Miss  NaN  22.020000
868          Mr  NaN  32.891990
878          Mr  NaN  32.891990
888        Miss  NaN  22.020000

[177 rows x 3 columns]

In [45]:
train[['name_title2','survived']].value_counts().sort_index()

name_title2  survived
Dr           0             4
             1             3
Master       0            17
             1            23
Miss         0            55
             1           131
Mr           0           447
             1            84
Mrs          0            26
             1           101
dtype: int64

In [46]:
train['age_gubun'] = train['age']//10
train['age_gubun']
train['age_gubun'] = train['age_gubun'].fillna(10)
train['age_gubun'] = train['age_gubun'].astype('int')

In [47]:
train[['age_gubun','survived']].value_counts().sort_index()

age_gubun  survived
0          0            24
           1            38
1          0            61
           1            41
2          0           143
           1            77
3          0            94
           1            73
4          0            55
           1            34
5          0            28
           1            20
6          0            13
           1             6
7          0             6
8          1             1
10         0           125
           1            52
dtype: int64

In [48]:
pd.crosstab(index = train['pclass'],columns = train['age_gubun'])

age_gubun  0   1    2   3   4   5   6   7   8    10
pclass                                             
1           3  18   34  50  37  27  13   3   1   30
2          17  18   53  48  18  15   3   1   0   11
3          42  66  133  69  34   6   3   2   0  136

In [49]:
pd.crosstab(index=train['pclass'] ,columns=train['age_gubun'] ,values=train['survived'] ,aggfunc='sum')

age_gubun    0     1     2     3     4     5    6    7    8     10
pclass                                                            
1           2.0  15.0  24.0  39.0  22.0  15.0  4.0  0.0  1.0  14.0
2          17.0   9.0  22.0  19.0  10.0   5.0  1.0  0.0  NaN   4.0
3          19.0  17.0  31.0  15.0   2.0   0.0  1.0  0.0  NaN  34.0

In [50]:
pd.crosstab(index=train['pclass'], columns=[train['age_gubun'],train['survived']])

age_gubun  0       1        2       3       4       5      6     7  8    10  \
survived    0   1   0   1    0   1   0   1   0   1   0   1  0  1  0  1    0   
pclass                                                                        
1           1   2   3  15   10  24  11  39  15  22  12  15  9  4  3  1   16   
2           0  17   9   9   31  22  29  19   8  10  10   5  2  1  1  0    7   
3          23  19  49  17  102  31  54  15  32   2   6   0  2  1  2  0  102   

age_gubun      
survived    1  
pclass         
1          14  
2           4  
3          34

In [51]:
train['family'] = train['sibsp'] + train['parch'] +1
train.head(10)

passengerid  survived  pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                name     sex   age  sibsp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   parch            ticket     fare cabin embarked cabin2 name_title  \
0      0         A/5 21171   7.2500   NaN        S    NaN         Mr   
1      0          PC 17599  71.2833   C85        C      C        Mrs   
2      0  STON/O2. 3101282   7.9250   NaN        S    NaN       Miss   
3      0            113803  53.1000  C123        S      C        Mrs   
4      0            373450   8.0500   NaN        S    NaN         Mr   
5      0            330877   8.4583   NaN        Q    NaN         Mr   
6      0             17463  51.8625   E46        S      E         Mr   
7      1            349909  21.0750   NaN        S    NaN     Master   
8      2            347742  11.1333   NaN        S    NaN        Mrs   
9      0            237736  30.0708   NaN        C    NaN        Mrs   

  name_title2      age2  age_gubun  family  
0          Mr  22.00000          2       2  
1         Mrs  38.00000          3       2  
2        Miss  26.00000          2       1  
3         Mrs  35.00000          3       2  
4          Mr  35.00000          3       1  
5          Mr  32.89199         10       1  
6          Mr  54.00000          5       1  
7      Master   2.00000          0       5  
8         Mrs  27.00000          2       3  
9         Mrs  14.00000          1       2

In [52]:
train[['family','survived']].value_counts().sort_index()

family  survived
1       0           374
        1           163
2       0            72
        1            89
3       0            43
        1            59
4       0             8
        1            21
5       0            12
        1             3
6       0            19
        1             3
7       0             8
        1             4
8       0             6
11      0             7
dtype: int64

In [54]:
train = train[['survived','pclass','sex','name_title2','age_gubun','family', 'age2']]
train.columns

Index(['survived', 'pclass', 'sex', 'name_title2', 'age_gubun', 'family',
       'age2'],
      dtype='object')

In [55]:
train.select_dtypes('object').columns

Index(['sex', 'name_title2'], dtype='object')

In [57]:
train.isna().sum()

survived       0
pclass         0
sex            0
name_title2    0
age_gubun      0
family         0
age2           0
dtype: int64

In [69]:
train.dtypes[train.dtypes=='object'].index

Index(['name_title2'], dtype='object')

In [61]:
train['sex'].unique()

array(['male', 'female'], dtype=object)

In [68]:
train['sex'] = train['sex'].replace(['male', 'female'],[1,0])
train['sex']

0      1
1      0
2      0
3      0
4      1
      ..
886    1
887    0
888    0
889    1
890    1
Name: sex, Length: 891, dtype: int64

In [74]:
train['names_title3'] = train['name_title2'].astype('category').cat.categories
train.head()

ValueError: Length of values (5) does not match length of index (891)